NOnion.Tests/CircuitHelper.cs

```diff
@@ -57,5 +63,34 @@ static async internal Task<List<CircuitNodeDetail>> GetRandomRoutersForDirectory
                     .Select(x => ConvertToCircuitNodeDetail(x))
                     .ToList();
         }
+
+        //FIXME: SLOW
+        static async internal Task<List<CircuitNodeDetail>> GetRandomRoutersForDirectoryBrowsingWithRetry(int count = 1)
+        {
+            Exception lastException = null;
+            var retry = 0;
+
+            while (retry < DirectoryAccessRetryLimit)
+            {
+                try
+                {
+                    return await GetRandomRoutersForDirectoryBrowsing(count);
+                }
+                catch (Exception ex)
+                {
+                    if (ex is CircuitDestroyedException || ex is GuardConnectionFailedException)
+                    {
+
+                        lastException = ex;
+                        retry++;
+                        continue;
+                    }
+
+                    throw;
+                }
+            }
+
+            throw lastException;
```

Comment:
and now that this method is much more clear and small, why not make it `while(true), and check if `retry < DirectoryAccessRetryLimit` inside the catch, this way you don't need to do `throw foo` but just `throw`, which preserves the exception better.